File: docs/beginner/module-02-basic-search/code-samples/csharp/02_PhraseSearch.cs
/*
Phrase Search - Module 2 C# Examples
Exact phrase matching in Azure AI Search using .NET SDK

This module demonstrates:
- Exact phrase search with quotes
- Comparing phrase vs individual terms
- Understanding when to use phrase search
- Phrase search best practices
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Models;

namespace AzureSearchHandbook.Module02.BasicSearch
{
    public class PhraseSearch
    {
        private readonly SearchClient _searchClient;

        public PhraseSearch(SearchClient searchClient)
        {
            _searchClient = searchClient ?? throw new ArgumentNullException(nameof(searchClient));
        }

        /// <summary>
        /// Search for an exact phrase using quotes
        /// </summary>
        /// <param name="phrase">Exact phrase to search for</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results</returns>
        public async Task<SearchResults<SearchDocument>> ExactPhraseSearchAsync(string phrase, int top = 10)
        {
            try
            {
                // Wrap phrase in quotes for exact matching
                var quotedPhrase = $"\"{phrase}\"";
                Console.WriteLine($"Performing exact phrase search: {quotedPhrase}");

                var searchOptions = new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                };

                var results = await _searchClient.SearchAsync<SearchDocument>(quotedPhrase, searchOptions);
                
                Console.WriteLine($"Found exact phrase matches");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in phrase search: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Search for individual terms (without quotes)
        /// </summary>
        /// <param name="phrase">Terms to search for individually</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results</returns>
        public async Task<SearchResults<SearchDocument>> IndividualTermsSearchAsync(string phrase, int top = 10)
        {
            try
            {
                Console.WriteLine($"Performing individual terms search: '{phrase}'");

                var searchOptions = new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                };

                var results = await _searchClient.SearchAsync<SearchDocument>(phrase, searchOptions);
                
                Console.WriteLine($"Found results for individual terms");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in individual terms search: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Compare exact phrase search vs individual terms search
        /// </summary>
        /// <param name="phrase">Phrase to compare</param>
        /// <param name="top">Maximum number of results for each search</param>
        /// <returns>Tuple of (phrase_results, terms_results)</returns>
        public async Task<(SearchResults<SearchDocument> phraseResults, SearchResults<SearchDocument> termsResults)> 
            ComparePhraseVsTermsAsync(string phrase, int top = 5)
        {
            var phraseResults = await ExactPhraseSearchAsync(phrase, top);
            var termsResults = await IndividualTermsSearchAsync(phrase, top);

            return (phraseResults, termsResults);
        }

        /// <summary>
        /// Display comparison between phrase and terms search
        /// </summary>
        /// <param name="phrase">Original phrase searched</param>
        /// <param name="phraseResults">Results from exact phrase search</param>
        /// <param name="termsResults">Results from individual terms search</param>
        public static void DisplayComparison(string phrase, 
            SearchResults<SearchDocument> phraseResults, 
            SearchResults<SearchDocument> termsResults)
        {
            Console.WriteLine($"\n🔤 Phrase vs Terms Comparison: '{phrase}'");
            Console.WriteLine(new string('=', 60));

            // Convert results to lists for easier handling
            var phraseList = phraseResults?.GetResults()?.ToList() ?? new List<SearchResult<SearchDocument>>();
            var termsList = termsResults?.GetResults()?.ToList() ?? new List<SearchResult<SearchDocument>>();

            // Exact phrase results
            Console.WriteLine($"\n1️⃣ EXACT PHRASE SEARCH: \"{phrase}\"");
            Console.WriteLine(new string('-', 40));
            Console.WriteLine($"Results found: {phraseList.Count}");

            if (phraseList.Any())
            {
                Console.WriteLine("Top matches:");
                for (int i = 0; i < Math.Min(3, phraseList.Count); i++)
                {
                    var result = phraseList[i];
                    var title = result.Document.TryGetValue("title", out var titleValue) ? 
                        titleValue?.ToString() : "No title";
                    var score = result.Score ?? 0.0;
                    Console.WriteLine($"  {i + 1}. {title} (Score: {score:F3})");
                }
            }
            else
            {
                Console.WriteLine("  No exact phrase matches found");
            }

            // Individual terms results
            Console.WriteLine($"\n2️⃣ INDIVIDUAL TERMS SEARCH: {phrase}");
            Console.WriteLine(new string('-', 40));
            Console.WriteLine($"Results found: {termsList.Count}");

            if (termsList.Any())
            {
                Console.WriteLine("Top matches:");
                for (int i = 0; i < Math.Min(3, termsList.Count); i++)
                {
                    var result = termsList[i];
                    var title = result.Document.TryGetValue("title", out var titleValue) ? 
                        titleValue?.ToString() : "No title";
                    var score = result.Score ?? 0.0;
                    Console.WriteLine($"  {i + 1}. {title} (Score: {score:F3})");
                }
            }
            else
            {
                Console.WriteLine("  No results found for individual terms");
            }

            // Analysis
            Console.WriteLine($"\n📊 COMPARISON ANALYSIS:");
            Console.WriteLine($"   Exact phrase: {phraseList.Count} results");
            Console.WriteLine($"   Individual terms: {termsList.Count} results");

            if (phraseList.Any() && termsList.Any())
            {
                var phraseAvg = phraseList.Where(r => r.Score.HasValue).Average(r => r.Score.Value);
                var termsAvg = termsList.Where(r => r.Score.HasValue).Average(r => r.Score.Value);
                Console.WriteLine($"   Average phrase score: {phraseAvg:F3}");
                Console.WriteLine($"   Average terms score: {termsAvg:F3}");
            }

            // Recommendations
            Console.WriteLine($"\n💡 RECOMMENDATIONS:");
            if (phraseList.Count > 0)
            {
                Console.WriteLine("   ✅ Exact phrase found - use phrase search for precision");
            }
            else if (termsList.Count > 0)
            {
                Console.WriteLine("   ⚠️ No exact phrase - individual terms provide broader results");
            }
            else
            {
                Console.WriteLine("   ❌ No results found - try different keywords or broader terms");
            }
        }

        /// <summary>
        /// Search for multiple phrases
        /// </summary>
        /// <param name="phrases">List of phrases to search for</param>
        /// <param name="top">Maximum results per phrase</param>
        /// <returns>Dictionary mapping phrases to their results</returns>
        public async Task<Dictionary<string, SearchResults<SearchDocument>>> MultiPhraseSearchAsync(
            IEnumerable<string> phrases, int top = 3)
        {
            var results = new Dictionary<string, SearchResults<SearchDocument>>();

            foreach (var phrase in phrases)
            {
                Console.WriteLine($"Searching for phrase: '{phrase}'");
                var phraseResults = await ExactPhraseSearchAsync(phrase, top);
                results[phrase] = phraseResults;
            }

            return results;
        }
    }

    /// <summary>
    /// Program class for demonstration
    /// </summary>
    public class PhraseSearchProgram
    {
        // Replace with your actual service details
        private const string ServiceEndpoint = "https://your-service.search.windows.net";
        private const string ApiKey = "your-api-key";
        private const string IndexName = "your-index-name";

        public static async Task Main(string[] args)
        {
            try
            {
                // Initialize search client
                var searchClient = new SearchClient(
                    new Uri(ServiceEndpoint),
                    IndexName,
                    new AzureKeyCredential(ApiKey)
                );

                await DemonstratePhraseSearchAsync(searchClient);
                await PhraseSearchBestPracticesAsync(searchClient);

                Console.WriteLine("\n💡 Next Steps:");
                Console.WriteLine("   - Try your own phrases with the examples above");
                Console.WriteLine("   - Compare results between phrase and terms search");
                Console.WriteLine("   - Check out 03_BooleanSearch.cs for combining terms with AND/OR");
                Console.WriteLine("   - Learn about wildcards in 04_WildcardSearch.cs");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ Demo failed: {ex.Message}");
                Console.WriteLine("Make sure your Azure AI Search service is configured correctly.");
            }
        }

        private static async Task DemonstratePhraseSearchAsync(SearchClient searchClient)
        {
            Console.WriteLine("🔤 Phrase Search Demonstration");
            Console.WriteLine(new string('=', 50));

            var searchOps = new PhraseSearch(searchClient);

            // Example 1: Compare phrase vs terms
            Console.WriteLine("\n1️⃣ Phrase vs Terms Comparison");

            var testPhrase = "machine learning";
            var (phraseResults, termsResults) = await searchOps.ComparePhraseVsTermsAsync(testPhrase, 5);
            PhraseSearch.DisplayComparison(testPhrase, phraseResults, termsResults);

            // Example 2: Another comparison
            Console.WriteLine($"\n{new string('=', 70)}");
            Console.WriteLine("\n2️⃣ Another Comparison Example");

            var testPhrase2 = "web development";
            var (phraseResults2, termsResults2) = await searchOps.ComparePhraseVsTermsAsync(testPhrase2, 5);
            PhraseSearch.DisplayComparison(testPhrase2, phraseResults2, termsResults2);

            // Example 3: Multiple phrase search
            Console.WriteLine($"\n{new string('=', 70)}");
            Console.WriteLine("\n3️⃣ Multiple Phrase Search");
            Console.WriteLine(new string('-', 30));

            var phrasesToSearch = new[]
            {
                "artificial intelligence",
                "data science",
                "software engineering"
            };

            var multiResults = await searchOps.MultiPhraseSearchAsync(phrasesToSearch, 2);

            foreach (var kvp in multiResults)
            {
                var phrase = kvp.Key;
                var results = kvp.Value.GetResults().ToList();

                Console.WriteLine($"\nPhrase: \"{phrase}\"");
                Console.WriteLine($"Results: {results.Count}");
                
                if (results.Any())
                {
                    var topResult = results.First();
                    var title = topResult.Document.TryGetValue("title", out var titleValue) ? 
                        titleValue?.ToString() : "No title";
                    var score = topResult.Score ?? 0.0;
                    Console.WriteLine($"Top match: {title} (Score: {score:F3})");
                }
                else
                {
                    Console.WriteLine("No matches found");
                }
            }

            Console.WriteLine("\n✅ Phrase search demonstration completed!");
        }

        private static async Task PhraseSearchBestPracticesAsync(SearchClient searchClient)
        {
            Console.WriteLine("\n📚 Phrase Search Best Practices");
            Console.WriteLine(new string('=', 50));

            var searchOps = new PhraseSearch(searchClient);

            Console.WriteLine("\n💡 When to Use Phrase Search:");
            Console.WriteLine("   ✅ Looking for specific technical terms");
            Console.WriteLine("   ✅ Searching for proper names or titles");
            Console.WriteLine("   ✅ Finding exact quotes or references");
            Console.WriteLine("   ✅ When word order matters");

            Console.WriteLine("\n⚠️ When NOT to Use Phrase Search:");
            Console.WriteLine("   ❌ General topic searches");
            Console.WriteLine("   ❌ When you want broader results");
            Console.WriteLine("   ❌ Searching for concepts (not exact terms)");
            Console.WriteLine("   ❌ When unsure of exact wording");

            // Demonstrate with examples
            Console.WriteLine("\n🧪 Practice Examples:");

            // Good phrase search examples
            var goodPhrases = new[]
            {
                "React hooks",
                "machine learning algorithm",
                "REST API"
            };

            Console.WriteLine("\n✅ Good Phrase Search Examples:");
            foreach (var phrase in goodPhrases)
            {
                var results = await searchOps.ExactPhraseSearchAsync(phrase, 1);
                var resultCount = results.GetResults().Count();
                var status = resultCount > 0 ? "Found" : "Not found";
                Console.WriteLine($"   \"{phrase}\": {status}");
            }

            // Show fallback strategy
            Console.WriteLine("\n🔄 Fallback Strategy Example:");
            var testPhrase = "deep learning neural networks";

            // Try exact phrase first
            var exactResults = await searchOps.ExactPhraseSearchAsync(testPhrase, 3);
            var exactCount = exactResults.GetResults().Count();
            Console.WriteLine($"   Exact phrase \"{testPhrase}\": {exactCount} results");

            if (exactCount == 0)
            {
                // Fallback to individual terms
                var termsResults = await searchOps.IndividualTermsSearchAsync(testPhrase, 3);
                var termsCount = termsResults.GetResults().Count();
                Console.WriteLine($"   Fallback to terms: {termsCount} results");

                if (termsCount > 0)
                {
                    Console.WriteLine("   💡 Recommendation: Use individual terms for broader results");
                }
            }
        }
    }
}

File: docs/beginner/module-02-basic-search/code-samples/javascript/02_phrase_search.js
/**
 * Phrase Search - Module 2 JavaScript Examples
 * Exact phrase matching in Azure AI Search using JavaScript SDK
 * 
 * This module demonstrates:
 * - Exact phrase search with quotes
 * - Comparing phrase vs individual terms
 * - Understanding when to use phrase search
 * - Phrase search best practices
 */

const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');

class PhraseSearch {
    /**
     * Initialize with a search client
     * @param {SearchClient} searchClient - Azure Search client
     */
    constructor(searchClient) {
        if (!searchClient) {
            throw new Error('SearchClient is required');
        }
        this.searchClient = searchClient;
    }

    /**
     * Search for an exact phrase using quotes
     * @param {string} phrase - Exact phrase to search for
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results
     */
    async exactPhraseSearch(phrase, top = 10) {
        try {
            // Wrap phrase in quotes for exact matching
            const quotedPhrase = `"${phrase}"`;
            console.log(`Performing exact phrase search: ${quotedPhrase}`);

            const searchOptions = {
                top: top,
                includeTotalCount: true
            };

            const results = await this.searchClient.search(quotedPhrase, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} exact phrase matches`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error in phrase search: ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Search for individual terms (without quotes)
     * @param {string} phrase - Terms to search for individually
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results
     */
    async individualTermsSearch(phrase, top = 10) {
        try {
            console.log(`Performing individual terms search: '${phrase}'`);

            const searchOptions = {
                top: top,
                includeTotalCount: true
            };

            const results = await this.searchClient.search(phrase, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} results for individual terms`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error in individual terms search: ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Compare exact phrase search vs individual terms search
     * @param {string} phrase - Phrase to compare
     * @param {number} top - Maximum number of results for each search
     * @returns {Promise<Object>} Object with phrase and terms results
     */
    async comparePhraseVsTerms(phrase, top = 5) {
        const phraseResults = await this.exactPhraseSearch(phrase, top);
        const termsResults = await this.individualTermsSearch(phrase, top);

        return { phraseResults, termsResults };
    }

    /**
     * Display comparison between phrase and terms search
     * @param {string} phrase - Original phrase searched
     * @param {Object} phraseResults - Results from exact phrase search
     * @param {Object} termsResults - Results from individual terms search
     */
    static displayComparison(phrase, phraseResults, termsResults) {
        console.log(`\n🔤 Phrase vs Terms Comparison: '${phrase}'`);
        console.log('='.repeat(60));

        // Exact phrase results
        console.log(`\n1️⃣ EXACT PHRASE SEARCH: "${phrase}"`);
        console.log('-'.repeat(40));
        console.log(`Results found: ${phraseResults.results.length}`);

        if (phraseResults.results.length > 0) {
            console.log('Top matches:');
            phraseResults.results.slice(0, 3).forEach((result, index) => {
                const title = result.document.title || 'No title';
                const score = result.score || 0.0;
                console.log(`  ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
            });
        } else {
            console.log('  No exact phrase matches found');
        }

        // Individual terms results
        console.log(`\n2️⃣ INDIVIDUAL TERMS SEARCH: ${phrase}`);
        console.log('-'.repeat(40));
        console.log(`Results found: ${termsResults.results.length}`);

        if (termsResults.results.length > 0) {
            console.log('Top matches:');
            termsResults.results.slice(0, 3).forEach((result, index) => {
                const title = result.document.title || 'No title';
                const score = result.score || 0.0;
                console.log(`  ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
            });
        } else {
            console.log('  No results found for individual terms');
        }

        // Analysis
        console.log(`\n📊 COMPARISON ANALYSIS:`);
        console.log(`   Exact phrase: ${phraseResults.results.length} results`);
        console.log(`   Individual terms: ${termsResults.results.length} results`);

        if (phraseResults.results.length > 0 && termsResults.results.length > 0) {
            const phraseAvg = phraseResults.results.reduce((sum, r) => sum + (r.score || 0), 0) / phraseResults.results.length;
            const termsAvg = termsResults.results.reduce((sum, r) => sum + (r.score || 0), 0) / termsResults.results.length;
            console.log(`   Average phrase score: ${phraseAvg.toFixed(3)}`);
            console.log(`   Average terms score: ${termsAvg.toFixed(3)}`);
        }

        // Recommendations
        console.log(`\n💡 RECOMMENDATIONS:`);
        if (phraseResults.results.length > 0) {
            console.log('   ✅ Exact phrase found - use phrase search for precision');
        } else if (termsResults.results.length > 0) {
            console.log('   ⚠️ No exact phrase - individual terms provide broader results');
        } else {
            console.log('   ❌ No results found - try different keywords or broader terms');
        }
    }

    /**
     * Search for multiple phrases
     * @param {Array<string>} phrases - List of phrases to search for
     * @param {number} top - Maximum results per phrase
     * @returns {Promise<Object>} Object mapping phrases to their results
     */
    async multiPhraseSearch(phrases, top = 3) {
        const results = {};

        for (const phrase of phrases) {
            console.log(`Searching for phrase: '${phrase}'`);
            const phraseResults = await this.exactPhraseSearch(phrase, top);
            results[phrase] = phraseResults;
        }

        return results;
    }
}

/**
 * Demonstrate phrase search operations
 */
async function demonstratePhraseSearch() {
    console.log('🔤 Phrase Search Demonstration');
    console.log('='.repeat(50));

    try {
        // Initialize search client (replace with your actual service details)
        const searchClient = new SearchClient(
            'https://your-service.search.windows.net',
            'your-index-name',
            new AzureKeyCredential('your-api-key')
        );

        const searchOps = new PhraseSearch(searchClient);

        // Example 1: Compare phrase vs terms
        console.log('\n1️⃣ Phrase vs Terms Comparison');

        const testPhrase = 'machine learning';
        const { phraseResults, termsResults } = await searchOps.comparePhraseVsTerms(testPhrase, 5);
        PhraseSearch.displayComparison(testPhrase, phraseResults, termsResults);

        // Example 2: Another comparison
        console.log('\n' + '='.repeat(70));
        console.log('\n2️⃣ Another Comparison Example');

        const testPhrase2 = 'web development';
        const { phraseResults: phraseResults2, termsResults: termsResults2 } = 
            await searchOps.comparePhraseVsTerms(testPhrase2, 5);
        PhraseSearch.displayComparison(testPhrase2, phraseResults2, termsResults2);

        // Example 3: Multiple phrase search
        console.log('\n' + '='.repeat(70));
        console.log('\n3️⃣ Multiple Phrase Search');
        console.log('-'.repeat(30));

        const phrasesToSearch = [
            'artificial intelligence',
            'data science',
            'software engineering'
        ];

        const multiResults = await searchOps.multiPhraseSearch(phrasesToSearch, 2);

        for (const [phrase, results] of Object.entries(multiResults)) {
            console.log(`\nPhrase: "${phrase}"`);
            console.log(`Results: ${results.results.length}`);
            
            if (results.results.length > 0) {
                const topResult = results.results[0];
                const title = topResult.document.title || 'No title';
                const score = topResult.score || 0.0;
                console.log(`Top match: ${title} (Score: ${score.toFixed(3)})`);
            } else {
                console.log('No matches found');
            }
        }

        console.log('\n✅ Phrase search demonstration completed!');

    } catch (error) {
        console.error(`❌ Demo failed: ${error.message}`);
        console.log('Make sure your Azure AI Search service is configured correctly.');
    }
}

/**
 * Demonstrate best practices for phrase search
 */
async function phraseSearchBestPractices() {
    console.log('\n📚 Phrase Search Best Practices');
    console.log('='.repeat(50));

    try {
        // Initialize search client (replace with your actual service details)
        const searchClient = new SearchClient(
            'https://your-service.search.windows.net',
            'your-index-name',
            new AzureKeyCredential('your-api-key')
        );

        const searchOps = new PhraseSearch(searchClient);

        console.log('\n💡 When to Use Phrase Search:');
        console.log('   ✅ Looking for specific technical terms');
        console.log('   ✅ Searching for proper names or titles');
        console.log('   ✅ Finding exact quotes or references');
        console.log('   ✅ When word order matters');

        console.log('\n⚠️ When NOT to Use Phrase Search:');
        console.log('   ❌ General topic searches');
        console.log('   ❌ When you want broader results');
        console.log('   ❌ Searching for concepts (not exact terms)');
        console.log('   ❌ When unsure of exact wording');

        // Demonstrate with examples
        console.log('\n🧪 Practice Examples:');

        // Good phrase search examples
        const goodPhrases = [
            'React hooks',
            'machine learning algorithm',
            'REST API'
        ];

        console.log('\n✅ Good Phrase Search Examples:');
        for (const phrase of goodPhrases) {
            const results = await searchOps.exactPhraseSearch(phrase, 1);
            const status = results.results.length > 0 ? 'Found' : 'Not found';
            console.log(`   "${phrase}": ${status}`);
        }

        // Show fallback strategy
        console.log('\n🔄 Fallback Strategy Example:');
        const testPhrase = 'deep learning neural networks';

        // Try exact phrase first
        const exactResults = await searchOps.exactPhraseSearch(testPhrase, 3);
        console.log(`   Exact phrase "${testPhrase}": ${exactResults.results.length} results`);

        if (exactResults.results.length === 0) {
            // Fallback to individual terms
            const termsResults = await searchOps.individualTermsSearch(testPhrase, 3);
            console.log(`   Fallback to terms: ${termsResults.results.length} results`);

            if (termsResults.results.length > 0) {
                console.log('   💡 Recommendation: Use individual terms for broader results');
            }
        }

    } catch (error) {
        console.error(`❌ Best practices demo failed: ${error.message}`);
    }
}

// Main execution
async function main() {
    try {
        await demonstratePhraseSearch();
        await phraseSearchBestPractices();

        console.log('\n💡 Next Steps:');
        console.log('   - Try your own phrases with the examples above');
        console.log('   - Compare results between phrase and terms search');
        console.log('   - Check out 03_boolean_search.js for combining terms with AND/OR');
        console.log('   - Learn about wildcards in 04_wildcard_search.js');
    } catch (error) {
        console.error(`❌ Main execution failed: ${error.message}`);
    }
}

// Export for use in other modules
module.exports = {
    PhraseSearch,
    demonstratePhraseSearch,
    phraseSearchBestPractices
};

// Run if this file is executed directly
if (require.main === module) {
    main();
}